torch/csrc/jit/passes/canonicalize.cpp

```diff
@@ -47,5 +48,220 @@ std::shared_ptr<Graph> Canonicalize(
   return r;
 }
 
+// Which index in b's owning Node is b
+size_t blockIndex(const Block* b) {
+  auto n = b->owningNode();
+  AT_ASSERT(n);
+  for (size_t i = 0; i < n->blocks().size(); ++i) {
+    if (n->blocks()[i] == b) {
+      return i;
+    }
+  }
+  AT_ASSERT(false);
+}
+
+size_t blocksFromGraphBlock(Node* n) {
+  size_t dist = 0;
+  while (n->owningBlock()->owningNode()) {
+    n = n->owningBlock()->owningNode();
+    ++dist;
+  }
+  return dist;
+}
+
+/*
+ * This establishes a canonical ordering of nodes.
+ * If n1 and n2 are in the same block, whichever node appears first
+ * is before the other.
+ * If n1 and n2 are contained in different blocks of an if node,
+ * then whichever block is in the true block is ordered before the other.
+ * If n1 contains n2, then n1 is before n2. This has the nice property that
+ * whichever node appears first in a dump of the graph is before the other.
+ * NB: this is not a topological index. Topologically, two nodes in
+ * different blocks of an if node are not topologically < or > each other.
+ */
+bool isBefore(Node* n1, Node* n2) {
+  // Invalid to call with the same node as both args
+  AT_ASSERT(n1 != n2);
+
+  // Set n1 and n2 to be the number of blocks from the Graph block
+  size_t d_1 = blocksFromGraphBlock(n1);
+  size_t d_2 = blocksFromGraphBlock(n2);
+
+  for (; d_1 > d_2; --d_1) {
+    n1 = n1->owningBlock()->owningNode();
+    // n2 contains n1
+    if (n1 == n2) {
+      return false;
+    }
+  }
+
+  for (; d_2 > d_1; --d_2) {
+    n2 = n2->owningBlock()->owningNode();
+    // n1 contains n2
+    if (n2 == n1) {
+      return true;
+    }
+  }
+
+  // Now they are the same numer of blocks from the graph block,
+  // recurse upwards, checking if they are on the same block
+  while (true) {
+    if (n1->owningBlock() == n2->owningBlock()) {
+      return n1->isBefore(n2);
+    }
+
+    auto new_n1 = n1->owningBlock()->owningNode();
+    auto new_n2 = n2->owningBlock()->owningNode();
+
+    AT_ASSERT(new_n1 != nullptr);
+    AT_ASSERT(new_n2 != nullptr);
+
+    if (new_n1 == new_n2) {
+      // take whichever node is in the earlier block
+      auto index_1 = blockIndex(n1->owningBlock());
+      auto index_2 = blockIndex(n2->owningBlock());
+      return index_1 < index_2;
+    }
+
+    n1 = new_n1;
+    n2 = new_n2;
+  }
+}
+
+std::vector<size_t> sort_indexes(at::ArrayRef<Value*> values) {
+  // initialize original index locations
+  std::vector<size_t> idx(values.size());
+  std::iota(idx.begin(), idx.end(), 0);
+
+  // Sort values based on canonical ordering of their first usage
+  std::sort(idx.begin(), idx.end(), [&values](size_t i1, size_t i2) {
+    if (values[i1]->uses().size() == 0 && values[i2]->uses().size() == 0) {
+      return i1 < i2;
+    }
+    if (values[i1]->uses().size() == 0) {
+      return false;
+    } else if (values[i2]->uses().size() == 0) {
+      return true;
+    }
+
+    auto fst_v1 = values[i1]->uses()[0];
```

Comment:
i think this is fine to use ? i think it's stable

Reply:
But it isn't the first use. It might take carefully constructing a test case to make this so, but there is nothing keeping them in order. That is only one step away from causing this pass to break under extremely hard-to-understand circumstances.

Reply:
Pretty sure this is a real bug. It hasn't been hit yet because nothing has transformed the graph enough to make the Uses get out of sync with the  order in the graph, but it is a bug waiting to happen.